ntopng: 2.0 -> 3.8

avdv opened the pull request:
Built without [nedge](https://www.ntop.org/products/traffic-analysis/ntopng-edge/).

* add readline80 for building `third-party/lua-5.3.5`
* add mariadb, which is required when building without nedge
* add libmaxminddb which is required according to `docs/README.compilation`
* fetch `nDPI` version 2.6, which is found and build automatically
* update patch 0001: adapt to code reformatting
* remove patch 0002: the callback directory must no longer be writable
* remove patch 0003: SOCKET is no longer declared

###### Motivation for this change

ntopng 3.8 was released almost a year ago. It makes 2/3 patches unnecessary.
 
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

avdv committed:
ntopng: 2.0 -> 3.8

Built without nedge.

* add readline80 for building `third-party/lua-5.3.5`
* add mariadb, which is required when building without nedge
* add libmaxminddb which is required according to `docs/README.compilation`
* fetch `nDPI` version 2.6, which is found and build automatically
* update patch 0001: adapt to code reformatting
* remove patch 0002: the callback directory must no longer be writable
* remove patch 0003: SOCKET is no longer declared

dtzWill commented:
Have you tried, or taken a look at, the ntopng nixos module (`./nixos/modules/services/networking/ntopng.nix`)?  Does it need any changes?

I ask because it mentions redis used for persistent storage, which maybe is something mariadb is used for?

Anyway someone should vet the nixos module before moving ahead with this.  Shouldn't be anything tricky but I can't say without bouncing into the code/changes/etc.
Maybe it's time for a ntopng nixos test to help? :grin:

c0bw3b commented on pkgs/tools/networking/ntopng/default.nix:
```diff
@@ -1,41 +1,44 @@
-  buildInputs = [ libpcap/* gnutls libgcrypt*/ libxml2 glib geoip geolite-legacy
-    sqlite which autoreconfHook git pkgconfig groff curl json_c luajit zeromq
+  buildInputs = [ libpcap/* gnutls libgcrypt*/ libxml2 libmaxminddb glib geoip geolite-legacy
+    sqlite which autoreconfHook git pkgconfig groff curl json_c luajit zeromq mariadb readline80
```
`autoreconfHook` and `pkgconfig` belong in `nativeBuildInputs`. Maybe `which` and `groff` too?
+ is `git` still necessary now that you fetch nDPI with fetchurl?

avdv replied:
> `autoreconfHook` and `pkgconfig` belong in `nativeBuildInputs`. Maybe `which` and `groff` too?

Changed.

> is git still necessary now that you fetch nDPI with nix fetchers?

The configure script calls git in order to determine the commit hash, but this fails anyway, expectedly:
```
fatal: not a git repository (or any parent up to mount point /)
```
I removed it, so now the error / warning is different:
```
./configure: line 3845: git: command not found
```

c0bw3b commented on pkgs/tools/networking/ntopng/default.nix:
```diff
@@ -1,41 +1,44 @@
-  name = "ntopng-2.0";
+let
+  ndpi = fetchurl {
+    url = "https://github.com/ntop/nDPI/archive/2.6.tar.gz";
```
GitHub-generated _archives_ are not stable over time. You should rather use `fetchzip` or `fetchFromGitHub` here.

avdv replied:
Too bad, but what do I do then? I mean, after fetching the files, say via `fetchFromGitHub`, I need to put it into the build directory, since it needs to be writable. Do I have to make it a (sub-)derivation then?

avdv replied:
After some back-and-forth, I managed to change it to use `fetchFromGitHub` instead.

c0bw3b commented on pkgs/tools/networking/ntopng/default.nix:
```diff
@@ -1,41 +1,44 @@
+  };
+in stdenv.mkDerivation rec {
+  version = "3.8";
+  name = "ntopng-${version}";
```
```suggestion
  pname = "ntopng";
```

`name` will be defined automatically

avdv commented:
> Have you tried, or taken a look at, the ntopng nixos module (`./nixos/modules/services/networking/ntopng.nix`)? Does it need any changes?

No, I have not tried it.

I am also still a bit confused about the ntopng and ntopng-edge separation and what it means; and why the former needs the mysql / mariasql client library, ie. why it is not simply optional.

> I ask because it mentions redis used for persistent storage, which maybe is something mariadb is used for?

Matter of fact, it does run without mariadb, just like it did before. It is only an option for dumping:
```
[--dump-flows|-F] <mode>            | Dump expired flows. Mode:
                                    | es            Dump in ElasticSearch database
                                    |   Format:
                                    |   es;<mapping type>;<idx name>;<es URL>;<http auth>
                                    |   Example:
                                    |   es;ntopng;ntopng-%Y.%m.%d;http://localhost:9200/_bulk;
                                    |   Notes:
                                    |   The <idx name> accepts the strftime() format.
                                    |   <mapping type>s have been removed starting at
                                    |   ElasticSearch version 6. <mapping type> values whill therefore be
                                    |   ignored when using versions greater than or equal to 6.
                                    |
                                    | logstash      Dump in LogStash engine
                                    |   Format:
                                    |   logstash;<host>;<proto>;<port>
                                    |   Example:
                                    |   logstash;localhost;tcp;5510
                                    |
                                    | mysql         Dump in MySQL database
                                    |   Format:
                                    |   mysql;<host[@port]|socket>;<dbname>;<table name>;<user>;<pw>
                                    |   mysql;localhost;ntopng;flows;root;
                                    |
                                    | mysql-nprobe  Read from an nProbe-generated MySQL database
                                    |   Format:
                                    |   mysql-nprobe;<host|socket>;<dbname>;<prefix>;<user>;<pw>
                                    |   mysql-nprobe;localhost;ntopng;nf;root;
                                    |   Notes:
                                    |    The <prefix> must be the same as used in nProbe.
                                    |    Only one ntopng -i <interface> is allowed.
                                    |    Flows are only read. Dump is assumed to be done by nProbe.
                                    |   Example:
                                    |     ./nprobe ... --mysql="localhost:ntopng:nf:root:root"
                                    |     ./ntopng ... --dump-flows="mysql-nprobe;localhost;ntopng;nf;root;root"
```

avdv committed:
Move `autoreconfHook`, `groff`... into `nativeBuildInputs`

* remove git from buildInputs

avdv committed:
ntopng: Let name be set automatically

Co-Authored-By: Renaud <c0bw3b@users.noreply.github.com>

avdv commented:
Oh, one more thing that needs to change is that the new daemon requires one to change the password. The default user/password is accepted, but you have to give a new one. Which does not work currently, since it probably cannot store the password anywhere...

avdv committed:
ntopng: Use `fetchFromGitHub` instead `fetchurl`

avdv commented:
> Oh, one more thing that needs to change is that the new daemon requires one to change the password. The default user/password is accepted, but you have to give a new one. Which does not work currently, since it probably cannot store the password anywhere...

Actually, password hashes are stored in the Redis. But somehow changing the password via the web interface does not work. I could only trick it by setting the password and the `ntopng.admin_password_changed` keys manually with redis-cli.

That way, I could see the main interface, but only for a few seconds - then I was redirected to the login page again. Seems we need this patch https://github.com/ntop/ntopng/commit/07702712187347ef448a14647d051c223755b4cd

avdv committed:
ntopng: Add patch to fix login issue

avdv committed:
ntopng: Remove needless substitutions

```
substituteStream(): WARNING: pattern '/bin/rm' doesn't match anything in file 'autogen.sh'
...
substituteStream(): WARNING: pattern '/usr/local' doesn't match anything in file 'src/Ntop.cpp'
```

avdv committed:
ntopng: Remove dependency on luajit

The project unconditionally uses the Lua version part of its source tree as
a third party module. They disabled the JIT compiler because of some memory issues.

lordcirth commented:
Testing on Ubuntu 18.04:

`
lordcirth@nezha:~/gitRepos/nixpkgs$ nix-build -A redis -o redis
lordcirth@nezha:~/gitRepos/nixpkgs$ nix-build -A ntopng -o ntopng
lordcirth@nezha:~/gitRepos/nixpkgs$ ./redis/bin/redis-server
lordcirth@nezha:~/gitRepos/nixpkgs$ mkdir /var/lib/ntopng
# I didn't want to create an ntopng user just for this
lordcirth@nezha:~/gitRepos/nixpkgs$ chmod 777 /var/lib/ntopng
lordcirth@nezha:~/gitRepos/nixpkgs$ sudo ./ntopng/bin/ntopng
`

Seems to work.

avdv committed:
ntopng: Use Lua 5.3 from nixpkgs

Instead of building liblua from the third-party folder, use the static
library from lua5_3 instead.

avdv commented:
I'll look into this again sooner or later. Also, version 4.0 has been released, which may be even better to integrate.

c0bw3b commented:
Superseded by #159538
